client/src/components/QA/AnswerModal.jsx

```diff
@@ -1,35 +1,170 @@
+    super(props);
+
+    this.state = {
+      photos: [],
+      questionId: this.props.questionId
+    };
+
+    this.submitAnswer = this.submitAnswer.bind(this);
+    this.uploadPhotos = this.uploadPhotos.bind(this);
+    this.clickHandler = this.clickHandler.bind(this);
+  }
+
+  clickHandler(e) {
+    this.props.handleTrackingClick(e, e.currentTarget.className, 'Questions & Answers');
+    this.props.openThumbnail(e);
+  }
+
+  uploadPhotos(e) {
+    let photos = [];
+
+    for (let key in e.target.files) {
+      if (key !== 'length' && key !== 'item') {
+        photos.push(URL.createObjectURL(e.target.files[key]));
+      }
+    }
+
+    if (e.target.files.length > 5) {
+      alert('You may only upload 5 images');
+      e.target.value = '';
+    } else {
+      this.setState({
+        photos
+      });
+    }
+  }
+
+  submitAnswer(e) {
+    e.preventDefault();
+
+    const answer = document.getElementById('modal-answer').value;
+    const name = document.getElementById('modal-answer-nickname').value;
+    const email = document.getElementById('modal-answer-email').value;
+    const photos = document.getElementById('modal-photos').files;
+
+    let photoFileNames = [];
+
+    for (let key in photos) {
+      if (key !== 'item' && key !== 'length') {
+        photoFileNames.push(photos[key].name);
+      }
+    }
+
+    let tracker = {
+      answer,
+      name,
+      email
+    };
+
+    const regexp = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+
+    const emailValid = regexp.test(email);
+
+    if (!answer || !name || !email || !emailValid) {
+      let string = 'You must enter or address the following:\n';
+
+      for (let key in tracker) {
+        if (!tracker[key]) {
+          string += `- ${key.slice(0, 1).toUpperCase() + key.slice(1)}\n`;
+        }
+      }
+
+      if (email && !emailValid) {
+        string += '- Your email must be formatted correctly\n';
+      }
+
+      if (!this.photosValid(photoFileNames)) {
+        string += '- You may only upload .jpg, .jpeg, .bmp, .gif, and .png files';
+      }
+
+      alert(string);
+      return;
+    }
+
+    let modal = document.querySelector('.modal');
+    modal.style.display = 'none';
+
+    let formElement = document.querySelector('#add-answer');
+    let formData = new FormData(formElement);
+    formData.append('questionId', this.props.questionId);
+
+
+    API.postAnswer(formData)
+      .then(response => {
+        document.getElementById('modal-answer').value = '';
+        document.getElementById('modal-answer-nickname').value = '';
+        document.getElementById('modal-answer-email').value = '';
+        document.getElementById('modal-photos').value = '';
+        this.setState({
+          photos: []
+        });
+        this.props.getQuestions();
+      })
+      .catch(err => {
+        console.log(err);
+      });
+  }
+
+  photosValid(photos) {
```

Comment:
I just started reading about validating file types. We should probably validate on the server. Multer has a built in `fileFilter` method. If we only check the extension, someone could change virus.exe to virus.jpg. I would hope that imagebb would also perform this validation though.

Reply:
good point! might leave the client validation as is for now and start looking into the `fileFilter` method for the server. i noticed the API itself rejects certain things too (e.g. badly formatted emails), i wonder if it does that with the image files.